### ISSUE TITLE
option to use case insensitive PKs for random row validation 

### ISSUE THREAD

**Issue by pankajsgit** [GoogleCloudPlatform/professional-services-data-validator#668]
The row validation function for DVT fails to generate the query when using case insensitive schemas

Source: SQL Server
Target: Postgres

Source table:
CREATE TABLE [Sales].[Customer](
	[CustomerID] [int] IDENTITY(1,1) NOT FOR REPLICATION NOT NULL,
	[PersonID] [int] NULL,
	[StoreID] [int] NULL,
	[TerritoryID] [int] NULL,
	[AccountNumber]  AS (isnull('AW'+[dbo].[ufnLeadingZeros]([CustomerID]),'')),
	[rowguid] [uniqueidentifier] ROWGUIDCOL  NOT NULL,
	[ModifiedDate] [datetime] NOT NULL,
 CONSTRAINT [PK_Customer_CustomerID] PRIMARY KEY CLUSTERED 
(
	[CustomerID] ASC
)WITH (PAD_INDEX = OFF, STATISTICS_NORECOMPUTE = OFF, IGNORE_DUP_KEY = OFF, ALLOW_ROW_LOCKS = ON, ALLOW_PAGE_LOCKS = ON, OPTIMIZE_FOR_SEQUENTIAL_KEY = OFF) ON [PRIMARY]
) ON [PRIMARY]

Target Table:
CREATE TABLE IF NOT EXISTS sales.customer
(
    customerid integer NOT NULL,
    personid integer,
    storeid integer,
    territoryid integer,
    accountnumber text COLLATE pg_catalog."default" NOT NULL,
    rowguid uuid NOT NULL DEFAULT uuid_generate_v4(),
    modifieddate timestamp with time zone NOT NULL DEFAULT CURRENT_TIMESTAMP,
    CONSTRAINT idx_22345_pk_customer_customerid PRIMARY KEY (customerid)
)

Row validation query:
 data-validation --verbose validate row -sc db_sqlserver -tc db_postgres  --primary-keys customerid -tbls sales.customer --hash 'StoreID'

It's failing with the following message:
ibis.common.exceptions.IbisTypeError: 'StoreID' is not a field in ['customerid', 'personid', 'storeid', 'territoryid', 'accountnumber', 'rowguid', 'modifieddate']

**Comment by nehanene15** [comment]
This is possible with YAML row validation, generated with the `-c config.yaml` flag. You can specify the different cases in source_column name and target_column_name.

**Comment by nehanene15** [comment]
While this has a workaround, we can explore getting a case-insensitive column name when select columns are specified in the -hash flag (not `-hash '*'`) similar to what we do in column validation.

**Comment by dominikhron** [comment]
Hi @nehanene15, I'm afraid the YAML row validation you mentioned as a workaround doesn't help. I have created two tables similar to @pankajsgit to test this on.

---

**Teradata source** table:

```sql
CREATE TABLE CZ_TST_NUCLEUS.DVT_TEST_DECIMAL
( 
	ROW_ID VARCHAR(15) NOT NULL,
	DECIMAL_COL DECIMAL(10,3) 
);

INSERT INTO CZ_TST_NUCLEUS.DVT_TEST_DECIMAL VALUES ( '1', 123.123 );
INSERT INTO CZ_TST_NUCLEUS.DVT_TEST_DECIMAL VALUES ( '2', 456.456 );
INSERT INTO CZ_TST_NUCLEUS.DVT_TEST_DECIMAL VALUES ( '3', 789.789 );
```

**Biguery target** table:

```sql
create table vfcz_dh_edw_adhoc_dev_01_s.dvt_test_decimal
( 
  row_id string not null,
  decimal_col numeric 
);

insert into vfcz_dh_edw_adhoc_dev_01_s.dvt_test_decimal VALUES
( '1', 123.123 ),
( '2', 456.456 ),
( '3', 789.789 );
```

---

First off, I ran the following **row validation**:

```
data-validation validate row -sc MY_TD_CONN -tc MY_BQ_CONN -tbls CZ_TST_NUCLEUS.DVT_TEST_DECIMAL=vfcz_dh_edw_adhoc_dev_01_s.dvt_test_decimal -pk row_id --hash 'row_id, decimal_col' -rr -rbs 3
```

The **result** is an error similar to the one reported in this issue:

```
ibis.common.exceptions.IbisTypeError: Column 'ROW_ID' is not found in table. Existing columns: 'hash__all', 'row_id'.
```

Next, I tried YAML row validation, generated with the `-c config.yaml` flag as you suggested and this already has the correct cases specified in the source_column name and target_column name. Please see the config generated in the attached  [config.txt](https://github.com/GoogleCloudPlatform/professional-services-data-validator/files/12586429/config.txt) file (GitHub doesn't support .yaml). I've tried all different case combinations in this file, but none have helped resolve the issue, I always get some permutation of the same error.

---

Interestingly, a similar **row validation without random row** sampling works just fine:

```
data-validation validate row -sc MY_TD_CONN -tc MY_BQ_CONN -tbls CZ_TST_NUCLEUS.DVT_TEST_DECIMAL=vfcz_dh_edw_adhoc_dev_01_s.dvt_test_decimal -pk row_id --hash 'row_id, decimal_col'
```

Here, the **result** is a successful row validation:

```
╒═══════════════════╤═══════════════════╤═════════════════════════════════╤══════════════════════╤══════════════════════════════════════════════════════════════════╤══════════════════════════════════════════════════════════════════╤══════════════════╤═════════════════════╤══════════════════════════════════════╕
│ validation_name   │ validation_type   │ source_table_name               │ source_column_name   │ source_agg_value                                                 │ target_agg_value                                                 │ pct_difference   │ validation_status   │ run_id                               │
╞═══════════════════╪═══════════════════╪═════════════════════════════════╪══════════════════════╪══════════════════════════════════════════════════════════════════╪══════════════════════════════════════════════════════════════════╪══════════════════╪═════════════════════╪══════════════════════════════════════╡
│ hash__all         │ Row               │ CZ_TST_NUCLEUS.DVT_TEST_DECIMAL │ hash__all            │ 4e07408562bedb8b60ce05c1decfe3ad16b72230967de01f640b7e4729b49fce │ 4e07408562bedb8b60ce05c1decfe3ad16b72230967de01f640b7e4729b49fce │                  │ success             │ 491b8afc-e7dc-4101-877a-9674b2e97033 │
├───────────────────┼───────────────────┼─────────────────────────────────┼──────────────────────┼──────────────────────────────────────────────────────────────────┼──────────────────────────────────────────────────────────────────┼──────────────────┼─────────────────────┼──────────────────────────────────────┤
│ hash__all         │ Row               │ CZ_TST_NUCLEUS.DVT_TEST_DECIMAL │ hash__all            │ d4735e3a265e16eee03f59718b9b5d03019c07d8b6c51f90da3a666eec13ab35 │ d4735e3a265e16eee03f59718b9b5d03019c07d8b6c51f90da3a666eec13ab35 │                  │ success             │ 491b8afc-e7dc-4101-877a-9674b2e97033 │
├───────────────────┼───────────────────┼─────────────────────────────────┼──────────────────────┼──────────────────────────────────────────────────────────────────┼──────────────────────────────────────────────────────────────────┼──────────────────┼─────────────────────┼──────────────────────────────────────┤
│ hash__all         │ Row               │ CZ_TST_NUCLEUS.DVT_TEST_DECIMAL │ hash__all            │ 6b86b273ff34fce19d6b804eff5a3f5747ada4eaa22f1d49c01e52ddb7875b4b │ 6b86b273ff34fce19d6b804eff5a3f5747ada4eaa22f1d49c01e52ddb7875b4b │                  │ success             │ 491b8afc-e7dc-4101-877a-9674b2e97033 │
╘═══════════════════╧═══════════════════╧═════════════════════════════════╧══════════════════════╧══════════════════════════════════════════════════════════════════╧══════════════════════════════════════════════════════════════════╧══════════════════╧═════════════════════╧══════════════════════════════════════╛
```

**Comment by nehanene15** [comment]
@dominikhron Yeah - the original issue here is actually resolved with #921 which supports case insensitive columns for row validation. This is why your example of row validation without random row works.

Random row validation adds a filter on the PK (`WHERE pk IN (x,y,z)`) which I believe is causing the issue. We can update this issue to tackle case-insensitive primary keys for the random row flag specifically
